Fetch radio schedules data on server

tochwill opened the pull request:
Resolves #5537

In order to render the radio schedule for a service on the server, we need to fetch the data upfront, before we render the react application. Here, we fetch that data in parallel with the page data required to render a front page with radio schedules.

**Overall change:**
This PR adds a call on the server to fetch radio schedule data for front pages that require it. If we get the data back from the API call, we merge it into the page data that is passed into the application. This allows us to pass this data into the radio schedules component in order to be rendered on the server. The component is still able to fetch the data it needs on the client if it has not been passed data.

**Code changes:**

- Adds a 'withRadioSchedule' function, that wraps the data fetching promise to fetch page data, and merges in the radio schedule data. The getInitialData function first checks to see if a service requires radio schedules on the front page, and if it does, it uses this function. Otherwise it fetches page data as previously.

- Adds an initialData prop to the radioSchedule container, which allows for data fetched on the server to be passed in and rendered.

- Pulls the radio schedule data out of pageData in the front page container, and passes it down to the radio schedule container.

- Updates tests and data processing utility functions.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
-  [x] This PR requires manual testing

Testing notes:
- Regression test the current front pages without radio schedules - they should be unaffected by this change.
- Pages with radio schedules should now have the radio schedule rendered on the server, rather than appearing in after page load.
- If the fetch on the server for radio schedules data fails, the page should render without radio schedules. We should then attempt to fetch the data on the client and render the radio schedules component once if this data is successfully fetched.
- The integration test previously written for radio schedules can now be run, as we are rendering on the server.

tochwill committed:
Fetch radio schedules data on server is required, and pass into page data

tochwill committed:
Name fetch page data default export

tochwill commented on src/app/routes/home/getInitialData/index.js:
```diff
@@ -14,13 +17,69 @@ const transformJson = pipe(
-export default async (path) => {
+export default async ({ path, service, variant = 'default' }) => {
+  // should we think of a resusable pattern for using config outside of our react app?
+  // such as const config = await getConfig(service);
```
```suggestion
  // such as const config = await getConfig(service, variant);
```

tochwill committed:
Update src/app/routes/home/getInitialData/index.js

tochwill commented on src/server/index.jsx:
```diff
@@ -260,7 +260,11 @@ server
           urlPath,
         );
-        const data = await route.getInitialData(url);
+        const data = await route.getInitialData({
```
This has been updated across the board in https://github.com/bbc/simorgh/pull/6109 - we can pull those changes into this branch (or pull out into a smaller separate PR) if we want to merge the radio schedules work first.

tochwill replied:
These changes are now in https://github.com/bbc/simorgh/pull/6171/files

benjifellows committed:
process data on server

hotinglok committed:
Updated processRadioSchedule to handle errors if data isn't found

hotinglok committed:
Merge branch 'latest' of github.com:bbc/simorgh into ssr-radio-schedules

hotinglok committed:
Updated tests to have different endpoints

Commit:
Use Promise.all() to fetch data.

Commit:
Add prop types.

tochwill committed:
Move radio schedule data fetching into reusable function and mock config in tests

tochwill committed:
clean up useEffect deps

tochwill committed:
Rename withRadioSchedules to withRadioSchedule

tochwill committed:
No need to spread transformJson

EinsteinNjoroge committed:
Merge branch 'latest' into ssr-radio-schedules

hotinglok committed:
Edited endpoint function & updated/added tests

EinsteinNjoroge committed:
Merge branch 'latest' into ssr-radio-schedules

benjifellows committed:
Canonical RadioSchedule added tests with and without initial data

itimirichard committed:
add tests for withRadioSchedule

tochwill commented:
Dependent on https://github.com/bbc/simorgh/pull/6171/files

tochwill committed:
remove comments

tochwill committed:
Increase bundle size limit by 3KB

tochwill commented on src/app/lib/config/services/getConfig/index.js:
```diff
@@ -0,0 +1,22 @@
+/* This function allows for a dynamic import of service config outside of
+the react application (for example, for server side data fetching). Within
+the react application, the config can be found in the ServiceContext. */
+const getConfig = async (service, variant = 'default') => {
```
I'm  a little uncertain about this - but this seems like the best way of being able to use the service config outside of the react application, which appears to be impossible at the moment? Using a normal static import of all of the config and pulling out the one we need would bloat the bundle massively. This means that we're importing the config for this service here in one way, and using a 'loadable' react context in the react application. This seems like a sensible option without a big refactor in the way be load config, but I'm very open to suggestions.

tochwill replied:
What if we move this so it's not co-located with the config files, but only in the routes, where we need to access config like this?

tochwill commented on src/app/routes/home/getInitialData/index.js:
```diff
@@ -14,8 +17,26 @@ const transformJson = pipe(
+    ['radioSchedule', 'onFrontPage'],
+    config,
+  );
+
```
One thing we are not doing here is respecting the value of the toggle for radio schedules for this particular environment. We could import the static toggles json file and check that, but we are moving towards an approach where toggles come from an external endpoint, and are configured within the react application. So as is, if radio schedules is toggled off for an environment, we would still fetch the data for them on the server (and not render the component due to the toggles checks within the react application).

tochwill commented on src/app/routes/utils/fetchPageData/index.js:
```diff
@@ -67,12 +67,11 @@ const handleError = (e) => {
-const fetchData = (pathname) => {
-  const url = getUrl(pathname);
-
+export const fetchData = (url) => {
```
We wanted to reuse all of the data fetching logic, but pass in an endpoint that was unrelated to the logic handled by `getUrl`. We should rearrange the tests for fetchPageData to test the fetchData function if we stick with this.

tochwill replied:
Taking another look at this, if we want to handle errors differently, which we probably do, then we don't want to reuse this function. Have added the data fetching code as a function into the withRadioSchedules file which means we can handle errors with a different constant, and ignore all of the 'error propagation' that we don't need for this call - will push up changes after a conversation tomorrow.

tochwill commented on src/app/routes/utils/withRadioSchedule/index.test.js:
```diff
@@ -0,0 +1,99 @@
+      expect(status).toBe(404);
+    });
+  });
+});
```
We have tested so far:

- Successful data fetches (so the promise resolves)
- Unsuccessful data fetches (i.e. - the promise has resolved but the status code is not 2xx or 3xx)) 

How about if we encounter an actual error when fetching radio schedules data, so the promise is rejected? I think `fetchData` with currently throw an error that isn't caught anywhere which will give us a 500. If we want to render the page without radio schedules if the page data is fetched and radio schedules fails, we should handle this.

tochwill replied:
I was wrong here - `fetchData` catches the error, so this will render the page without radio schedules if the promise is rejected. I'm adding a test case.

tochwill commented on src/app/routes/utils/withRadioSchedule/index.js:
```diff
@@ -0,0 +1,35 @@
+    baseUrl: SIMORGH_BASE_URL,
+  });
+
+  const radioSchedulePromise = fetchData(radioScheduleUrl);
```
If we want to still render the page if this promise _rejects_ we can wrap this in error handling that resolves with no data, which will allow Promise.all to resolve if the page data is successfully fetched.

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Update integration snapshots

EinsteinNjoroge committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Update tests

tochwill committed:
Merge branch 'ssr-radio-schedules' of github.com:bbc/simorgh into ssr-radio-schedules

tochwill committed:
Pass variant to getInitialData in App

tochwill commented on src/integration/pages/articlePage/__snapshots__/mundo.test.js.snap:
```diff
@@ -1985,7 +1985,7 @@ exports[`Given I am on a Mundo Article page When I view the source code in the b
             defer
     >
-      [21]
+      [19]
```
I'm not sure why these numbers are being updated in these snapshots for the integration tests. Is it something to do with webpack chunking differently between test runs?

tochwill committed:
add test to cover radio schedules fetch promise rejecting

tochwill committed:
Destructure radio schedules json

mmcculloch commented on src/app/containers/App/App.test.jsx:
```diff
@@ -185,6 +185,7 @@ describe('App', () => {
           expect(route.getInitialData).toHaveBeenCalledWith({
             path: pathname,
             service: 'news',
```
Should we change this service to one that has an actual variant?

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -104,6 +104,9 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
 
   const { timeOnServer } = useContext(RequestContext);
+
+  const [schedule, setRadioSchedule] = useState(initialData);
```
Rename one of these so they match. e.g. `radioSchedule, setRadioSchedule`

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -120,24 +123,26 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
+      const fetchRadioScheduleData = (pathname) =>
+        fetch(pathname, { mode: 'no-cors' })
+          .then(handleResponse)
+          .catch((e) => logger.error(`HTTP Error: "${e}"`));
```
Update this to the new error format https://github.com/bbc/simorgh/blob/latest/src/app/lib/README.md#logging

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -120,24 +123,26 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
-  }, [endpoint, locale, script, service, timeOnServer, timezone]);
+    if (!schedule) {
+      const handleResponse = async (response) => {
+        const radioScheduleData = await response.json();
```
Check `response.ok` is `true` first, otherwise throw an error

mmcculloch replied:
We can also look into extra logging if status codes are not what we expect (200)

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
-  it('renders correctly for a service with a radio schedule and page frequency URL', async () => {
-    fetch.mockResponse(JSON.stringify(arabicRadioScheduleData));
+  describe('With initial data', () => {
+    it('renders correctly for a service with initial data', async () => {
```
Can just be `renders correctly for a service`

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
+        />,
+      );
+
+      await wait(() => {
```
Consider replacing with `act`

mmcculloch commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
+    });
+
+    it('does not render when data contains no programs', async () => {
+      fetch.mockResponse(
```
Consider testing for fetch failures as well.

mmcculloch commented on src/app/containers/RadioSchedule/index.jsx:
```diff
@@ -6,9 +6,13 @@ import { ServiceContext } from '#contexts/ServiceContext';
 import { RequestContext } from '#contexts/RequestContext';
 import useToggle from '#hooks/useToggle';
```
Double check the toggle is tested.

mmcculloch commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
 };
 
 export default (radioScheduleData, service, currentTime) => {
+  if (!radioScheduleData) {
```
Consider renaming to `data` and cleaning up var names

mmcculloch commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
+    return null;
+  }
+
   // finding latest program, that may or may not still be live. this is because there isn't
```
This comment needs to be moved now we think

mmcculloch commented on src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
+
   // finding latest program, that may or may not still be live. this is because there isn't
   // always a live program, in which case we show the most recently played program on demand.
+  const radioSchedules = pathOr([], ['schedules'], radioScheduleData);
```
`const radioSchedules = radioScheduleData.schedules || [];` Could also work here?

mmcculloch commented on src/app/pages/FrontPage/index.test.jsx:
```diff
@@ -34,7 +34,10 @@ let pageData;
   fetch.mockResponse(JSON.stringify(frontPageDataPidgin));
 
-  const response = await getInitialData('some-front-page-path');
+  const response = await getInitialData({
```
Check other occurrences of `getInitialData` to ensure an object is passed. Seems to not fail with the old string format.

thekp replied:
> Seems to not fail with the old string format.

I feel like this test still passes because of what `fetch.mockResponse` is doing. It would return a response regardless of what the fetch path is and in the case where the path is not defined in the object, I assume it's `undefined` but the response is returned anyways. 

This should be updated in this issue: https://github.com/bbc/simorgh/issues/5960 to be URL sensitive when fetch mocking.

mmcculloch commented on src/app/routes/home/getInitialData/index.js:
```diff
@@ -14,8 +17,26 @@ const transformJson = pipe(
-export default async ({ path }) => {
-  const { json, ...rest } = await fetchPageData(path);
+export default async ({ path, service, variant }) => {
+  const config = await getConfig(service, variant);
+
+  const hasRadioSchedule = pathOr(
+    false,
+    ['radioSchedule', 'hasRadioSchedule'],
+    config,
+  );
+
+  const radioScheduleOnFrontPage = pathOr(
+    false,
+    ['radioSchedule', 'onFrontPage'],
+    config,
+  );
+
+  const pageHasRadioSchedule = hasRadioSchedule && radioScheduleOnFrontPage;
```
Move this into a separate function `pageHasRadioSchedule(service, variant)`

mmcculloch commented on src/app/routes/home/getInitialData/index.test.js:
```diff
@@ -1,15 +1,92 @@
-  );
-  expect(pageData.promo.name).toEqual('Domot');
-  expect(pageData.content.groups.length).toBeTruthy();
+  it('should return essential data for a page without radio schedules to render', async () => {
```
Maybe reword to "data" rather than "essential data"

mmcculloch commented on src/app/routes/home/getInitialData/index.test.js:
```diff
@@ -1,15 +1,92 @@
+  });
+
+  it('should return data for service with radio schedules, but without radio schedules on front page', async () => {
+    getConfig.mockImplementationOnce(() => ({
```
A missing combination that could be tested
```
    getConfig.mockImplementationOnce(() => ({
      radioSchedule: {
        hasRadioSchedule: false,
        onFrontPage: true,
      },
    }));
```

mmcculloch commented on src/app/routes/utils/fetchPageData/index.js:
```diff
@@ -67,12 +67,11 @@ const handleError = (e) => {
-  const url = getUrl(pathname);
-
+export const fetchData = (url) => {
   logger.info(DATA_REQUEST_RECEIVED, { url });
```
Can we add another error constant?

mmcculloch commented on src/app/routes/utils/withRadioSchedule/index.test.js:
```diff
@@ -0,0 +1,120 @@
+    });
+
+    describe('fetch API promises rejected', () => {
+      it('should return page data without radio schedules if radio schedule fetch promise is rejected', async () => {
```
Would be good to test that this failure is being logged

mmcculloch commented on src/app/routes/utils/withRadioSchedule/index.test.js:
```diff
@@ -0,0 +1,120 @@
+      );
+
+      expect(json).toBeUndefined();
+      expect(status).toBe(404);
```
Would be good to test that this failure is being logged

mmcculloch commented on src/app/routes/utils/withRadioSchedule/index.test.js:
```diff
@@ -0,0 +1,120 @@
+      );
+
+      expect(json).toBeUndefined();
+      expect(status).toBe(404);
```
Would be good to test that this failure is being logged

mmcculloch commented on src/app/routes/utils/withRadioSchedule/index.js:
```diff
@@ -0,0 +1,35 @@
+import { getQueryString } from '#lib/utilities/urlParser';
+import processRadioSchedule from '#containers/RadioSchedule/utilities/processRadioSchedule';
+
+const withRadioSchedules = async (pageDataPromise, service, path) => {
```
Should be `withRadioSchedule` to match the file name

tochwill commented on src/app/lib/config/services/getConfig/index.js:
```diff
@@ -0,0 +1,22 @@
+    serviceConfig = config;
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(
```
Replace with the logger if we go with this approach.

tochwill committed:
Implemented changes from code review

tochwill committed:
Test canonical component in canonical radio schedule test

tochwill commented on src/app/routes/utils/withRadioSchedule/index.test.js:
```diff
@@ -0,0 +1,140 @@
+import fetchMock from 'fetch-mock';
+import loggerMock from '#testHelpers/loggerMock'; // Must be imported before fetchPageData
```
```suggestion
import loggerMock from '#testHelpers/loggerMock';
```

tochwill committed:
Update src/app/routes/utils/withRadioSchedule/index.test.js

tochwill committed:
Revert changes to fetchData, handle radio schedule data fetching in withRadioSchedule

tochwill committed:
Merge branch 'ssr-radio-schedules' of github.com:bbc/simorgh into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill commented:
I've gone through and implemented changes for all of the points raised in the swarm review (and updated a couple more things that I noticed going back through the PR). I have _not_ yet made changes for the following, which could be handled today or in follow up tickets.

src/app/containers/App/App.test.jsx
Should we change this service to one that has an actual variant?

src/app/containers/RadioSchedule/Canonical/index.jsx
We can also look into extra logging if status codes are not what we expect (200)

src/app/pages/FrontPage/index.test.jsx
Check other occurrences of getInitialData to ensure an object is passed. Seems to not fail with the old string format.

src/app/routes/utils/fetchPageData/index.js
Can we add another error constant? - note: I have stopped trying to reuse this function, and instead have included error handling in the `withRadioSchedule` function. We can add a new constant and replace the usage of `DATA_FETCH_ERROR`.

src/app/containers/RadioSchedule/Canonical/index.test.jsx
Consider replacing with act

This is covered by this ticket: https://github.com/bbc/simorgh/issues/6219

tochwill committed:
Fix snapshot

tochwill committed:
Update logging to correct format

tochwill committed:
Update bundle size

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Move dynamic config import to route

tochwill committed:
Update snapshots and add logging on schedule data fetch

tochwill committed:
Merge branch 'ssr-radio-schedules' of github.com:bbc/simorgh into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Update ramda defaults

EinsteinNjoroge committed:
Merge branch 'latest' into ssr-radio-schedules

oluoluoxenfree commented:
@tochwill Could you fill in more of the PR description? :)

thekp commented on src/app/routes/home/getInitialData/index.js:
```diff
@@ -14,8 +17,30 @@ const transformJson = pipe(
+  const pageHasRadioSchedule = await hasRadioSchedule(service, variant);
+
+  const { json, ...rest } = pageHasRadioSchedule
+    ? await withRadioSchedule(fetchPageData(path), service, path)
```
Is it worth changing this to accept an object?

tochwill replied:
Done!

thekp reviewed (commented):
Looks good to me overall. Snapshots seem to be failing integration tests 👀  and a small comment about the `withRadioSchedule` function.

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill commented:
Fixing merge conflicts

tochwill committed:
Obey updated linter

tochwill committed:
Update snapshot test

chris-hinds commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -91,55 +93,76 @@ const RadioFrequencyLink = styled.a`
+            logger.error(
+              JSON.stringify(
+                {
+                  event: DATA_FETCH_ERROR,
```
would it be better to have a custom event for radio schedule data? it would make it easier to filter through the logs?

tochwill replied:
Sure - I added that to do in another issue, but I can make the change as part of this PR instead.

tochwill replied:
Done! I will add to sumologic when I've made all requested changes and merged.

ryanmccombe commented on src/app/containers/RadioSchedule/Canonical/index.jsx:
```diff
@@ -91,55 +93,76 @@ const RadioFrequencyLink = styled.a`
+
+  const header = pathOr(null, ['header'], radioScheduleConfig);
   const frequenciesPageUrl = pathOr(
     null,
```
There is a simpler `path` function for when we don't need to specify a fallback value: https://ramdajs.com/docs/#path

`path` falls back to undefined - I don't think `null` vs `undefined` makes a difference in most instances here

tochwill replied:
I've just been looking at this too - I'll update the usage across this PR.

tochwill replied:
Actually as all of these examples are just one level deep into the object, I should be fine to default `radioScheduleConfig` to an empty object, and then destructure these values out.

tochwill replied:
Done! Have updated usages across this PR.

itimirichard commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -1,55 +1,173 @@
+      );
+      await matchSnapshotAsync(
+        <RadioScheduleWithContext
+          service="arabic"
+          initialData={initialData}
+          radioScheduleToggle
```
Do we need the `service` and `radioScheduleToggle` props here and in the other instances of `RadioScheduleWithContext` below? They don't seem to be needed within the component on line 16 above

tochwill replied:
We don't! Will remove them.

tochwill replied:
Done!

tochwill committed:
Use path instead of pathOr where possible

tochwill committed:
Remove incorrect props from radio schedule test

tochwill committed:
Update radio schedule logging

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

DenisHdz commented on src/app/containers/RadioSchedule/Canonical/index.test.jsx:
```diff
@@ -1,55 +1,153 @@
-      <RadioSchedulesWithContext service="arabic" radioScheduleToggle />,
-    );
-  });
+      await wait(() => {
```
Looks good to me overall 💯
Wondering if we still have to replace this with `await act( async () => {`

tochwill replied:
I would like to, I've added that to another ticket to do once this is reviewed, there's quite a lot of changes in here already!

tochwill replied:
https://github.com/bbc/simorgh/issues/6219

DenisHdz replied:
Excellent 👍

itimirichard reviewed (approved):
Thanks for making all these changes Tom. Looks good to me!

DenisHdz committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

tochwill committed:
Merge branch 'latest' into ssr-radio-schedules

paruchurisilpa committed:
Merge branch 'latest' into ssr-radio-schedules

paruchurisilpa committed:
Merge branch 'latest' into ssr-radio-schedules

paruchurisilpa commented:
* Regression testing on Radio schedules didnt show any issues.
   *  If the fetch on the server for radio schedules data fails, the page should render without radio schedules. We should then attempt to fetch the data on the client and render the radio schedules component once if this data is successfully fetched  --> This I have checked with help from Tom.
   * Integration tests have passed.

LGTM...

tochwill commented:
Thank you @paruchurisilpa :) I just noticed that the integration test is still commented out, I have amended it slightly and will push up to this branch before merging this in.

tochwill committed:
Enable radio schedule integration test

tochwill committed:
Merge branch 'ssr-radio-schedules' of github.com:bbc/simorgh into ssr-radio-schedules

tochwill committed:
guard against undefined error
